File: python/questions/subsets.py
from typing import List

class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        ln = len(nums)
        ans = [[]]
        tmp = []
        def bs(p: int) -> None:
            tmp.append(nums[p])
            ans.append(tmp.copy())
            for i in range(p+1, ln):
                bs(i)
            tmp.pop()
        for i in range(ln):
            bs(i)
        return ans

File: cpp/questions/subsets.cpp
#include <functional>
#include <vector>

using namespace std;

class Solution {
public:
    vector<vector<int>> subsets(vector<int>& nums) {
        int ns = nums.size();
        std::vector<std::vector<int>> ans = {{}};        
        std::vector<int> tmp;
        tmp.reserve(ns);
        std::function<void(int)> bs = [&](int p) -> void {
            tmp.emplace_back(nums[p]);
            ans.emplace_back(tmp);
            for (int i=p+1; i<ns; ++i) {
                bs(i);
            }
            tmp.pop_back();
        };
        for (int i=0; i<ns; ++i) {
            bs(i);
        }
        return ans;
    }
};
